compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -92,6 +92,15 @@ impl<'a> Parser<'a> {
             // FIXME(Centril): Can we unconditionally `allow_plus`?
             let ty = self.parse_ty_common(allow_plus, recover_qpath, AllowCVariadic::No)?;
             FnRetTy::Ty(ty)
+        } else if self.token == token::FatArrow {
+            // Don't `eat` to prevent `=>` from being added as an expected token which isn't
+            // actually expected and could only confuse users
+            self.bump();
+            self.struct_span_err(self.prev_token.span, "return types are denoted using `->`")
```

Comment:
I'd suggest that you add a `recover_fat_arrow` argument and only perform this recovery when the caller requests it - in particular, where `parse_ret_ty` is used in parsing path segments, this recovery shouldn't happen.

I expect that this example, which currently compiles, would no longer parse with the change as it exists currently (but would with my prior suggestion):

```rust
fn bar(_: u32) {}

fn baz() -> *const dyn Fn(u32) { unimplemented!() }

fn foo() {
    match () {
        _ if baz() == &bar as &dyn Fn(u32) => (),
        () => (),
    }
}
```

Please add the above as a test too.